[Builder] Expose GetImage interface for builder

dnephin opened the pull request:
Implement the second task in #32904

This PR removes the previous 3 Image functions from the builder.Backend interface and replaces them with a single function. Also refactors the interfaces between `imageMount`, `imageContects`, and `dispatchers.from`

dnephin committed:
Expose a smaller interface for the Builder retrieving images from daemon

Removes 3 methods from the builder.Backend interface
Remove the coupling between imageContexts, imageMounts and the builder.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Refactor interaction between dispatcher.from and dispatchState

Signed-off-by: Daniel Nephin <dnephin@docker.com>

tonistiigi commented:
Design LGTM

dnephin commented:
I think I fixed the test failures. The problem was that releaseableLayer was being returned with a closer over the original image/imageID. This meant that updates to the base layer were not reflected in the layer.

The change was to add an imageID to `ReleasableLayer.Mount()`. This works, but it makes the interface a bit weird. Now the first call returns an object that really has nothing to do with the image that it returns, it's more of an uninitialized object that happens to have a reference to the daemon.

My concern is that it wouldn't really be possible to model this with a remote API. We might want to consider a separate `MountImage()` like we had better. Will this interface make more sense later? WDYT?

tonistiigi commented:
As discussed privately the need for `Mount()` parameter should go away and `releaseableLayer` should be always backed by `layer.Layer`. Putting back to design-review until that is resolved.

dnephin committed:
Refactor imageContexts into two different structs.

buildStages now tracks the imageID and runConfig for a build stage

imageMounter tracks image mounts so they can released when the build ends.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
What do you mean by "backed by layer.Layer" ?  There is a RWLayer underlying the releaseableLayer

tonistiigi commented:
@dnephin Every instance of `releasableLayer` needs to be an instance of(or keep a reference to) `layer.Layer` that is the rootfs of the image returned. That makes sure that the data is protected until builder runs. When `Mount()` for that layer is called, `RWLayer` is created on top of that `layer.Layer` and mountpoint for that is made available. `releasableLayer.Release()` calls `layerstore.Release(layer)` in the implementation.

tonistiigi commented on daemon/build.go:
```diff
@@ -0,0 +1,125 @@
+
+func (rl *releaseableLayer) Release() error {
+	if rl.rwLayer == nil {
+		return nil
```
`roLayer` still needs to be released. Might be simpler to add a ref to `roLayer` in `releaseableLayer` instead of defining these callback methods.

tonistiigi replied:
For extra safety, maybe makes sense to define a finalizer for this type.

dnephin replied:
Oops, right. Thanks

tonistiigi commented on daemon/build.go:
```diff
@@ -0,0 +1,125 @@
+
+	mountFunc := func() (layer.RWLayer, error) {
+		mountID := stringid.GenerateRandomID()
+		return daemon.layerStore.CreateRWLayer(mountID, img.RootFS.ChainID(), nil)
```
`roLayer.ChainID` is bit safer

dnephin commented:
The build is green!

I believe I've addressed the issues with roLayer in releaseableLayer as well as the bug described in #33090

Read for another design review

dnephin commented on daemon/build.go:
```diff
@@ -0,0 +1,119 @@
+func newReleasableLayerForImage(img *image.Image, layerStore layer.Store) (*releaseableLayer, error) {
+	if img.RootFS.ChainID() == "" {
+		return &releaseableLayer{}, nil
+	}
```
When an image is created from scratch and doesn't include any copy/add this chainID is empty. I'll add a comment about this with the next round of changes after design review.

I guess ill also need to add a check to `Mount()` and provide a better error when roLayer is nil

tonistiigi replied:
Yes, `Mount()` seems unprotected atm. Other way would be return `nil`(or not call `GetImage` for scratch) as there is no point for keeping a reference for scratch.

dnephin replied:
`GetImage()` is never called with a scratch object. But if you do `FROM someimage` and that image doesn't have any `COPY/ADD`, then this `ChainID()` is still the empty string.

So we have to check for that condition here. We could either return a nop implementation of ReleasableLayer, but I decided against that because many of these nil checks are required anyway. So i thought it was be easier to just add a few more instead of creating another struct.

tonistiigi replied:
What about just returning `nil`. It could be better to detect this case here than to call `Mount()` and get the error. Also if we add `Exec()` then this takes layer as a parameter, so it would be better to detect this case before instead of passing the empty struct.

dnephin replied:
Oh ya, that is a lot better. The case where layer is nil is already handled.

I'll leave the nil checks in mount anyway for safety

tonistiigi commented on builder/dockerfile/imagecontext.go:
```diff
@@ -5,87 +5,141 @@ import (
+	s.sequence[len(s.sequence)-1].update(imageID, runConfig)
+}
+
+type getAndMountFunc func(string) (builder.Image, builder.ReleaseableLayer, error)
```
nit: what's the benefit of callback instead of using the interface dependency

dnephin replied:
The advantage is that we don't have to store all the opts that are only used in this one place. We only need them to construct the object. By storing just the function type that relationship is clear. If we were to store all the opts it wouldn't be clear.

tonistiigi reviewed (commented):
Just some comments, overall looks good

dnephin commented on builder/builder.go:
```diff
@@ -62,15 +58,11 @@ type Backend interface {
-	RunConfig() *container.Config
+// ImageBackend are the interface methods required from an image component
+type ImageBackend interface {
+	GetImageAndLayer(ctx context.Context, refOrID string, opts backend.GetImageAndLayerOptions) (Image, ReleaseableLayer, error)
```
It can't be `GetImage` because it conflicts with another daemon method already called `GetImage`.

I called it Layer, since the second return is named ReleaseableLayer. Should we rename that object as well? I think they should match.

Or should I make this `GetImageAndReleasableLayer()` ?

dnephin replied:
Ok, I'll use `GetImageAndReleasableLayer()` for now. It's easy to rename later when we have the ability to use snapshots. I think it will be confusing to readers if we introduce the term snapshot right  now.

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -235,24 +250,69 @@ func (b *Builder) getFromImage(dispatchState *dispatchState, shlex *ShellLex, na
-		return nil, nil
+		return scratchImage, nil
+	}
+	imageMount, err := b.imageMounter.Get(name)
```
That's true, but the purpose of the struct is still to manage the unmounts required because of the API we are using. I don't know what else to call it

tonistiigi commented on builder/dockerfile/dispatchers.go:
```diff
@@ -160,23 +161,34 @@ func dispatchCopy(req dispatchRequest) error {
+	if stage != nil {
+		imageRefOrID = stage.ImageID()
+	}
+	return b.imageMounter.Get(imageRefOrID)
```
re: previous comment
It isn't really that it manages unmounts. It more managers keeping the layer references. Unmount currently also happens on `release` but that will probably be updated. `imageRefs`, `images`, `imageSources`, `imageCache` ?

dnephin replied:
I think we already have `imageCache` so I'll go with imageSources

dnephin commented:
I've made those changes, and CI seems to be good except for some flaky tests

tonistiigi reviewed (approved):
LGTM

vdemeester reviewed (approved):
LGTM 🦁
